### PR TITLE
Support do-block syntax for redirect_std*

### PR DESCRIPTION
**Pull request by timholy**
Ref https://github.com/JuliaLang/julia/pull/18165

### CODE REVIEW

**Comment by timholy** [comment]
I don't have a copy of julia 0.3 lying around; anyone have any idea about the error? I tried adding a `@compat` but that didn't seem to help.

**Comment by TotalVerb** [comment]
v0.3 does not support defining a function without methods.

File: src/redirect.jl [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+Run the function `f` while redirecting `STDERR` to `stream`. Upon completion, `STDERR` is restored to its prior setting.
+"""
+redirect_stderr(f::Function, stream)
+
+"""
+    redirect_stdin(f::Function, stream)
+
+Run the function `f` while redirecting `STDIN` to `stream`. Upon completion, `STDIN` is restored to its prior setting.
+"""
+redirect_stdin(f::Function, stream)
```
Comment (stevengj) [inline src/redirect.jl]: It seems like overkill to have docstrings in Compat.jl.    We don't do that for any other functions that we copy from newer versions.   

And if you delete the docstrings, you can just put the code directly into `src/Compat.jl`.   And it should fix things on 0.3.

**Comment by timholy** [comment]
Not sure I know what you mean. This is just adding a method to each of the existing `redirect_std*` functions.

Or do you mean it's the docstring that's likely the problem?

**Comment by timholy** [comment]
Sorry, browser refresh delay left things out of sync. I'll delete the docstrings.

**Comment by timholy** [comment]
Hmm, still no dice. I worry it's because of the `Base.$F` part, but I've tried with and without `@compat` and it doesn't seem to matter.

File: src/Compat.jl [hunk 2]
```diff
@@ -1452,4 +1452,20 @@ else
     end
 end
 
+if VERSION < v"0.6.0-dev.374"
+    for (F,S) in ((:redirect_stdin, :STDIN), (:redirect_stdout, :STDOUT), (:redirect_stderr, :STDERR))
+        @eval @compat function Base.$F(f::Function, stream)
```
Comment (yuyichao) [inline src/Compat.jl]: `function Base.$F() ...` doesn't parse on 0.3. You can use `import` (note that `@eval import Base: $F` won't work on 0.3 either so you need to either use `Expr(:import)` or just list the functions manually outside the loop)

Comment (stevengj) [inline src/Compat.jl]: I would just import the functions outside the loop.

Comment (timholy) [inline src/Compat.jl]: If it's a parsing thing, that explains why `@compat` doesn't help.

File: src/Compat.jl [hunk 3]
```diff
@@ -1452,4 +1452,21 @@ else
...
 
+import Base: redirect_stdin, redirect_stdout, redirect_stderr
+if VERSION < v"0.6.0-dev.374"
+    for (F,S) in ((:redirect_stdin, :STDIN), (:redirect_stdout, :STDOUT), (:redirect_stderr, :STDERR))
+        @eval function $F(f::Function, stream)
+            STDOLD = $S
+            local ret
+            $F(stream)
+            try
+                ret = f()
```
Comment (TotalVerb) [inline src/Compat.jl]: Could save a few lines with `try f() finally $F(STDOLD) end`.

Comment (tkelman) [inline src/Compat.jl]: it should return the value of `f()` not the value of `$F` though, shouldn't it?

Comment (TotalVerb) [inline src/Compat.jl]: A `try` block's value is the tail statement in `try` or `catch` position, not the value in `finally` position.

**Comment by timholy** [comment]
Something killed the OSX builds, but it passed on all 3 versions on Linux and Windows.

**Comment by timholy** [comment]
Oh, and thanks for the help, folks!

**Pull request merged**